### PR TITLE
UI counter JS

### PR DESCRIPTION
**Pull request by Amstutz**
This PR proposes to introduce a slim interface to manipulate Counters through a standardized interface defined in [counter.js](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-bcea3fb2f8b062aee035d27c5154dc3a). This interface is already required by the existing Meta Bar and will be essential for the upcoming Notifications in the Meta Bar. 

This PR Contains the following:
 - Interface, and Implementation: The [interface](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-bcea3fb2f8b062aee035d27c5154dc3a) itself.  
 - Example: An [example](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-170b469dfba907015e1026525dff6156) showcasing the interface at work:
image
 - Refactoring: Some slight [Refactoring](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-5fc652ec00c9107799a2da9f24e40cce) of the Meta bar Using the new interface. 
 - JS _Units Tests_: Note, client side (JS) testings are highly experimental and not yet required. Further in is not sure, in which shape client side testing in future will be shaped out in ILIAS Development. See [Tests Readme](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-52bc354f156c7d4b36bd178f5d19de21) for more information. Note tests are not intended to the shaping of how JS testing will look like in future. They have been created out of pure necessity since writing JS code without any option of continuously testing is extremely painful.

### CODE REVIEW

**Comment by Amstutz** [comment]
Tests will be fixed ASAP <-- done

File: src/UI/templates/js/Counter/counter.js [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+				console.assert(this.hasStatusObject(),"Novelty "+MISSING_COUNTERTYPE_EXCEPTION_MSG);
+
+				return this.incrementStatusCount(this.getNoveltyCount()).setNoveltyTo(0);
+			};
+
+			/**
+			 * Interface returned by this function for public use (see return statement bellow)
+			 * The contained functions are implemented bellow
+			 */
+			var public_object_interace = {
```
Comment (klees) [inline src/UI/templates/js/Counter/counter.js:240]:
```suggestion
			var public_object_interface = {
```

File: src/UI/templates/js/Counter/counter.js [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+				//Therefore we generate a separate counter instance
+				$parent.each(function() {
+					//Note that *this* in this anonymous function points to the
+					//object in counter jquery object currently looped
+					var single_counter = getCounterObject($(this));
+					workload(single_counter,$(this));
+				});
+			};
+
+			return public_object_interace;
```
Comment (klees) [inline src/UI/templates/js/Counter/counter.js:340]:
```suggestion
			return public_object_interface;
```

File: src/UI/Implementation/Component/Counter/Renderer.php [hunk 3]
```diff
@@ -18,13 +18,25 @@ public function render(Component\Component $component, RendererInterface $defaul
...
     }
 
+    /**
+     * @inheritdoc
+     */
+    public function registerResources(\ILIAS\UI\Implementation\Render\ResourceRegistry $registry)
+    {
+        parent::registerResources($registry);
+        $registry->register('./src/UI/templates/js/Counter/counter.js');
+
```
Comment (klees) [inline src/UI/Implementation/Component/Counter/Renderer.php:37]:
```suggestion
```

File: src/UI/templates/js/Counter/counter.js [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+				 */
+				setNoveltyTo: this.setNoveltyTo,
+				/**
+				 * Sets all Status Counters inside a given counter object to a set value
+				 */
+				setStatusTo: this.setStatusTo,
+				/**
+				 * Increments all Novelty Counters inside a counter Object by some given value
+				 */
+				incrementNoveltyCount: this.incrementNoveltyCount,
```
Comment (mjansenDatabay) [inline src/UI/templates/js/Counter/counter.js:268]: Suggestion for a better name  `incrementNoveltyCountBy`?

**Comment by matthiaskunkel** [comment]
Jour Fixe, 04 NOV 2019 : We highly appreciate this suggestion and accept it for trunk. Please merge after Richard's review.

File: src/UI/templates/js/Counter/counter.js [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+				/**
+				 * Decrements all Novelty Counters inside a counter Object by some given value
+				 */
+				decrementNoveltyCount: this.decrementNoveltyCount,
```
Comment (mjansenDatabay) [inline src/UI/templates/js/Counter/counter.js:272]: Same as above.

File: src/UI/templates/js/Counter/counter.js [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+				/**
+				 * Increments all Status Counters inside a counter Object by some given value
+				 */
+				incrementStatusCount: this.incrementStatusCount,
```
Comment (mjansenDatabay) [inline src/UI/templates/js/Counter/counter.js:276]: Same as above.

File: src/UI/templates/js/Counter/counter.js [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+				/**
+				 * Decrements all Status Counters inside a counter Object by some given value
+				 */
+				decrementStatusCount: this.decrementStatusCount,
```
Comment (mjansenDatabay) [inline src/UI/templates/js/Counter/counter.js:280]: Same as above.

**Comment by gvollbach** [comment]
@Amstutz 
First of all, great to see that someone is using Javascript Unittests. 
But wouldn't it be better/easier to use some Javascript Testframework, like e.g. [Jasmine](https://jasmine.github.io/) or similar?

**Comment by Amstutz** [comment]
@gvollbach : Easier, no. That code took maybe 15 minutes to write. Better, definitely! Important, for everybody reading/seeing this. I am not advocating that THIS will be the way we should write our tests in future. In no way is this any good or complete.

Important things missing:
 - Usable output for some CI.
 - Good easy to use assertions
 - Easily trackable traces of the error
 - Nice messages
 - Good inclusion in some dev. Environment.

I worked with Jasmine in the past. I believe it is good and might be a solution for us. However, I also believe that we should evaluate such a solution in collabaration with each other and evaluate carefully. This choice IMO does have implication on the way we do write JS. However, this is not the time for this (few weeks before the release). Therefore this solution here, might help as intermediate solution to mitigate the gravest pain of not being able to write tests for JS in the very present.

**Comment by gvollbach** [comment]
@Amstutz 

>  I believe it is good and might be a solution for us. However, I also believe that we should evaluate such a solution in collabaration with each other and evaluate carefully... Therefore this solution here, might help as intermediate solution to mitigate the gravest pain of not being able to write tests for JS in the very present.

I totally agree with you on this point, thanks for your clarification.

**Review by klees — CHANGES_REQUESTED** [review]
Hi @Amstutz!

As discussed via skype, we most probably will need to revise this at some point when we know more about client side JS patterns. Until then, this is absolutely fine for me. This definitely looks clean and approachable, so I have no doubt that we won't have problems to migrate this to whatever pattern or approach we use in the future.

Please consider the following suggestions:

* [x] Be a little more verbose in the examples. E.g. create the counter object and then use it in a second statement instead of creating and using it with method chaining in one line. IMO this will make the examples more comprehensible, especially [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-170b469dfba907015e1026525dff6156R62).

Please implement the following changes:

* [x] I may be extremely wrong here (I'm not very good in JS, especially when it comes to `this`), but from my understanding of how `this` works, the usage in [`generateCounterObject`](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-bcea3fb2f8b062aee035d27c5154dc3aR67) is problematic. AFAIK, `this` binds to the last object that the function was invoked on (besides [some other rules](https://medium.com/quick-code/understanding-the-this-keyword-in-javascript-cb76d4c7c5e8)). Going up from the only [usage of the function in `getCounterObject`](https://medium.com/quick-code/understanding-the-this-keyword-in-javascript-cb76d4c7c5e8) to the [usage of `getCounterObject` in your example](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-bcea3fb2f8b062aee035d27c5154dc3aR19) `this` will be bound to `il.UI.counter`. But from how I understand your code, you expect `this` to be bound to a new object when used. Since this is not the case, you will in fact only be able to use one counter object in parallel and get spooky distant interactions otherwise. I think you could fix this by using `return new generateCounterObject` [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2292/files#diff-bcea3fb2f8b062aee035d27c5154dc3aR45) As I said, this might be incorrect, semantics of `this` are tough...

Thanks a lot for being so thorough here. I really appreciate the tests in JS.

Best regards!

**Comment by Amstutz** [comment]
hi @klees

Concerning the example
---
Concerning the example *this* was supposed to be a little riddle for the interested viewer ;-). However, you are right it might just spook a potential client. Should be better now ;-)

Concerning *this*:
---
### TL/DR: Thanks for pointing out the issue. this is now properly bound with the latest commit.

Extended Version: Thanks a lot for putting your finger on this. Loving JS just for the brain teasers it is causing now and then ;-). Since we want to collaboratively improve our JS here some extended thought on the topic and the reasoning behind not following your suggestion of just using *new*.

You say, you are no expert, but you definitely have the right sensors to detect smelly code. There was definitely a smell around the usage of *this*, caused by some shit in the middle of it. As discussed, I refactored from extending JQuery to just using the JQuery object internally. However for utter laziness I left some uses of this meandering around in the code. It does not any harm, but is bad nevertheless. Some explanation (no JS expert either, so double check everything think along, this is just my current understanding of the matter):

**The new keyword JS**: When I understand everything correctly, is new in JS just some glorified binder, binding an empty object to some function and returning the new object. Note that this will returned by the new statement after binding it to the empty object if no return value is given by default. So, return this is not required. However, to show case the thing, here some example code:

```
//Generating the object by using new.
var myObjectWithNew = new someObjectConstructor("param1","param2");
myObjectWithNew.someFunction(); //Will return "Values: param1,param2,1"
myObjectWithNew.someFunction(); //Will return "Values: param1,param2,2"
var myObjectWithNew2 = new someObjectConstructor("param1","param2");
myObjectWithNew2.someFunction(); //Will return "Values: param1,param2,1"

//Does exactly the same as new here. Bind is used. {} is the new object we bind this in someObjectConstructor and return it. Note that return here is required, to get a handle on the object.
var newBoundObject = someObjectConstructor.bind({},"param1","param2")();
newBoundObject.someFunction(); //Will return "Values: param1,param2,1"
newBoundObject.someFunction(); //Will return "Values: param1,param2,2"
var newBoundObject2 = someObjectConstructor.bind({},"param1","param2")();
newBoundObject2.someFunction(); //Will again return "Values: param1,param2,1"

//Note that bind is more flexible, giving us the opportunity to e.g. construct a curried function
var curriedObjectCunstructor = someObjectConstructor.bind({},"param1","param2");
curriedObjectCunstructor("param2").someFunction(); //Will return "Values: param1,param2,1"

//To further clarify what bind does here an example with a none empty bind object.
var newBoundObject = someObjectConstructor.bind({someOtherValue: "value"},"param1","param2")();
newBoundObject.someFunction(); //Will return "Values: param1,param2,1"
newBoundObject.someOtherFunction(); //Will return "value"
```

So, why all this fuss here?

**Well let’s take a closer look the *this***: As @klees correctly points out, the behaviour of this is somewhat hard to follow, if not properly bound (as in this smelly context here). In our case the usage of this binds directly to the global window (IMO not counter, since we never bound this to counter), which is the shit the smell is has been bound to. However, there are not side effects, since we only bind the functions to *this*. The JQuery object is only used internally in the closure by var, so this would not have any consequences. However the code is misleading and a future refactoring could easily fall into the non-transparent trap by trying to make use of *this*. However, I do not want to make use of *new*. By using new, IMO a promise is given, that we return *this*, in the end of the constructor by implicitly returning nothing. However, this is not the case. A closure containing the interface is returned in order to keep things private. Therefore the following change is IMO more transparent:

Change:
`return generateCounterObject($counter);`
To:
```
var CounterObjectConstructor = generateCounterObject.bind({});
return CounterObjectConstructor($counter);
```

Oh I just loved the little „unit tests“ while doing this slight refactoring ;-)

**Comment by Amstutz** [comment]
May I merge?
